src/optimize.rs

```diff
@@ -0,0 +1,1049 @@
+        locals: &mut Vec<(u32, wasm_encoder::ValType)>,
+        num_locals: &mut u32,
+        new_insts: &mut Vec<CowInst>,
+        temp_callee_local: u32,
+        call_site_index: u32,
+        table_index: u32,
+        type_index: u32,
+    ) -> Result<Option<StackEntry<'a>>> {
+        // If we haven't already reached our maximum inlining depth...
+        if (on_stack.len() - 1) >= self.max_inline_depth {
+            return Ok(None);
+        }
+
+        // ... and we have profiling information for this call site...
+        let call_site = match context.profile.call_sites.get(&call_site_index) {
+            Some(x) => x,
+            None => return Ok(None),
+        };
+
+        // ... and this call site is hot enough...
+        if call_site.total_call_count < self.min_total_calls {
+            return Ok(None);
+        }
+
+        // ... then get the hottest callee table index...
+        let callee = call_site
+            .callee_to_count
+            .iter()
+            .map(|(callee, count)| (*callee, *count))
+            .max_by(|a, b| a.1.cmp(&b.1).then(b.0.cmp(&a.0)));
+        let (callee_index_in_table, callee_count) = match callee {
+            Some(x) => x,
+            None => return Ok(None),
+        };
+
+        // ... and if that hottest callee is called often enough...
+        let callee_ratio = callee_count as f64 / call_site.total_call_count as f64;
+        if callee_ratio < self.min_ratio {
+            return Ok(None);
+        }
+
+        // ... and if we statically know what function index `table[callee]`
+        // is...
+        let callee_func_index = match context.tables[table_index].get(callee_index_in_table) {
+            Some(x) => x,
+            None => return Ok(None),
+        };
+
+        // ... and if that function index is not already on our winlining stack
+        // (i.e. we aren't in a recursive inlining chain)...
+        if on_stack.contains(&callee_func_index) {
+            return Ok(None);
+        }
+
+        // ... and if that function has the correct type (if this is not true
+        // then either the profile is bogus/mismatched or every time the call
+        // site was executed it trapped)...
+        if context.funcs[usize::try_from(callee_func_index - context.num_imported_funcs).unwrap()]
+            != type_index
+        {
+            return Ok(None);
+        }
+
+        // ... then we can winline this callee and push it onto our stack!
+
+        let defined_func_index = callee_func_index - context.num_imported_funcs;
+        let locals_delta = *num_locals;
+        let func_body = context.func_bodies[usize::try_from(defined_func_index).unwrap()].clone();
+        let ops = func_body
+            .get_operators_reader()?
+            .into_iter_with_offsets()
+            .peekable();
+
+        // Add the first half of our speculative inlining sequence, before the
+        // inlined body:
+        //
+        //     local.tee $temp_callee_local
```

Comment:
We noted earlier when going over this code that `$temp_callee_local` might be clobbered in the case that we do nested inlining, but actually I'm not sure I see the problem anymore -- its live-range here extends from this tee to the compare below in the case that we do match, and otherwise to the local.get below, but in the latter case, the inlined body isn't executed; in other words, it's never live across the inlined code. So I think this is actually OK? (Can we add a comment here either way describing the way we're able to share a single local?)